sympy/stats/stochastic_process_types.py

```diff
@@ -676,31 +766,761 @@ def _transient2absorbing(self):
+        >>> X.canonical_form()
+        ([1, 2, 3, 4, 5], Matrix([
+        [  1,   0,   0,   0,   0],
+        [  0, 1/2, 1/2,   0,   0],
+        [2/5, 2/5, 1/5,   0,   0],
+        [1/2,   0,   0, 1/2,   0],
+        [  0, 1/2,   0,   0, 1/2]]))
+
+        The output should be the same as the first call but setting a state space
+        results in that state space being sorted instead of kept in place.
+
+        See Also
+        ========
+        sympy.stats.DiscreteMarkovChain.decompose
+        """
+        trans_probs = self.transition_probabilities
+        n = self.num_states
+        if (trans_probs is None) or (n is None):
+            return None
+
+        if isinstance(trans_probs, MatrixSymbol) or isinstance(n, Symbol):
+            raise NotImplementedError("The transient and recurrent states cannot be determined.")
+
+        classes = self.communication_classes()
+        r_states = []
+        t_states = []
+        for c in classes:
+            if c.recurrent:
+                r_states += c.states
+            else:
+                t_states += c.states
+
+        states = r_states + t_states
+        P_new = Matrix(trans_probs)
+        for row in range(n):
+            for col in range(n):
+                P_new[row, col] = trans_probs[states[row], states[col]]
+
+        # convert states to the user's states
+        states = [self.state_space.args[state] for state in states]
+        return states, P_new
+
+    def decompose(self):
+        """
+        The transition matrix can be decomposed into 4 submatrices:
+
+        - A - the submatrix from recurrent states to recurrent states
+        - B - the submatrix from transient to recurrent states
+        - C - the submatrix from transient to transient states
+        - 0 - the submatrix of zeros for recurrent to transient states
+
+        Returns
+        =======
+        (states, A, B, C)
+
+            ``states`` - a list of state names with the first being
+            the recurrent states and the last being
+            the transient states in the order
+            of the row names of A and then the row names of C.
+            ``A`` - the submatrix from recurrent states to recurrent states.
+            ``B`` - the submatrix from transient to recurrent states.
+            ``C`` - the submatrix from transient to transient states.
+
+        Examples
+        ========
+        >>> from sympy.stats import DiscreteMarkovChain
+        >>> from sympy import Matrix, S
+
+        You can decompose this matrix for example
+
+        >>> T = Matrix([[S(1)/2, S(1)/2, 0, 0, 0],
+        ...             [S(2)/5, S(1)/5, S(2)/5, 0, 0],
+        ...             [0, 0, 1, 0, 0],
+        ...             [0, 0, S(1)/2, S(1)/2, 0],
+        ...             [S(1)/2, 0, 0, 0, S(1)/2]])
+        >>> X = DiscreteMarkovChain('X', trans_probs=T)
+        >>> X.decompose()
+        ([2, 0, 1, 3, 4], Matrix([[1]]), Matrix([
```

Comment:
This kind of example would be clearer if you extract the parts and print them separately like
```
>>> states, A, B, C = X.decompose()
>>> states
[2, 0, 1, 3, 4]
>>> A   # recurrent to recurrent
...
```